Source/AwsCommonRuntimeKit/mqtt/Mqtt5Client.swift

```diff
@@ -259,43 +313,69 @@ public class Mqtt5Client {
-                guard result == AWS_OP_SUCCESS else {
-                    continuationCore.release()
-                    return continuation.resume(throwing: CommonRunTimeError.crtError(CRTError.makeFromLastError()))
+                self.rwlock.read {
+                    // validate the client in case close() is called.
+                    guard let rawValue = self.rawValue else {
+                        continuationCore.release()
+                        return continuation.resume(throwing: CommonRunTimeError.crtError(CRTError.makeFromLastError()))
+                    }
+                    let result = aws_mqtt5_client_unsubscribe(rawValue, unsubscribePacketPointer, &callbackOptions)
+                    guard result == AWS_OP_SUCCESS else {
+                        continuationCore.release()
+                        return continuation.resume(throwing: CommonRunTimeError.crtError(CRTError.makeFromLastError()))
+                    }
                 }
             }
         }
     }
 
+    /// Discard all operations and cleanup the client. It is MANDATORY function to call to release the client.
     public func close() {
-        self.callbackCore.close()
-        aws_mqtt5_client_release(rawValue)
-        rawValue = nil
+        self.rwlock.write {
+            self.callbackFlag = false
+            aws_mqtt5_client_release(rawValue)
+            self.rawValue = nil
+        }
+    }
+
+    /////////////////////////////////////////////////////////
+    // helper functions for self retained reference
+    private func callbackUserData() -> UnsafeMutableRawPointer {
+        return Unmanaged<Mqtt5Client>.passRetained(self).toOpaque()
+    }
+
+    private func release() {
+        self.rwlock.write {
+            self.callbackFlag = false
+        }
+        Unmanaged<Mqtt5Client>.passUnretained(self).release()
     }
+
 }
 
 // MARK: - Internal/Private
 
 /// Handles lifecycle events from native Mqtt Client
 internal func MqttClientHandleLifecycleEvent(_ lifecycleEvent: UnsafePointer<aws_mqtt5_client_lifecycle_event>?) {
 
-    guard let lifecycleEvent: UnsafePointer<aws_mqtt5_client_lifecycle_event> = lifecycleEvent else {
+    guard let lifecycleEvent: UnsafePointer<aws_mqtt5_client_lifecycle_event> = lifecycleEvent,
+        let userData = lifecycleEvent.pointee.user_data else {
         fatalError("MqttClientLifecycleEvents was called from native without an aws_mqtt5_client_lifecycle_event.")
     }
-    let callbackCore = Unmanaged<MqttCallbackCore>.fromOpaque(lifecycleEvent.pointee.user_data).takeUnretainedValue()
+    let client = Unmanaged<Mqtt5Client>.fromOpaque(userData).takeUnretainedValue()
     let crtError = CRTError(code: lifecycleEvent.pointee.error_code)
 
     // validate the callback flag, if flag is false, return
-    callbackCore.rwlock.read {
-        if callbackCore.callbackFlag == false { return }
+    client.rwlock.read {
```

Comment:
Trivial: Should we move `let client` and `let crtError` into the client lock because if close has been called, I think trying to check the callbackFlag might be accessing potentially bad memory? Why are we using the flag check here instead of doing a nil check on assigning client?

Reply:
The lock is used to protect the flag and native client. It is safe to access the swift client here. Meanwhile, the swift client would always available until the termination callback get called.